examples/sui-move/escrow/sources/shared_example.move

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// An escrow for atomic swap of objects using shared objects without a trusted third party.
+/// Swap via shared Escrow 
+/// 
+/// 1. A user creates a shared object requesting for a specific object ID
+/// and posting up their object ID. 
+/// 
+/// 2. The other user puts in their object which has the same object ID
+/// and the exchange is fired in the same txn.
+module escrow::shared_example {
+    use std::option::{Self, Option};
+
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    /// An object held in escrow
+    struct Exchange<T: key + store> has key, store {
+        id: UID,
+        /// owner of the escrowed object
+        sender: address,
+        /// intended recipient of the escrowed object
+        recipient: address,
+        /// ID of the object `sender` wants in exchange
+        // TODO: this is probably a bad idea if the object is mutable.
+        // that can be fixed by asking for an additional approval
+        // from `sender`, but let's keep it simple for now.
```

Comment:
Can you employ a similar fix to what I did in `example.move` using `Locked<T>` and `Key`?  It will work quite similarly in the shared case, but it's asymmetric:

- Both parties `lock` their objects to get a `Locked<T>` and a `Key`.
- One party creates the `Exchange` shared object using their `Locked<T>` and `Key` and indicating the `ID` for the key of the object they want in exchange.
- The other party initiates the `swap` by indicating the `Exchange` object they want to participate in, as well as providing their `Locked<T>` and `Key`.
- The `swap` checks that the key passed in to the swap matches the one in the `Exchange` object, and that it unlocks the `Locked<T>` as well, and if all is well, it takes the offered object in `Exchange` and sends it to the transaction sender, and sends the locked object (now unlocked) to the address that did the offering.

Additionally you may want to factor out `Locked<T>`, `Key` and the associated functions into their own module.

Finally, cc @tzakian, because this is an ideal example where shared object deletion could really improve things.  Instead of storing an `Option<T>`, you could store just a `T` and the act of either revoking or finalising the swap would also delete the `Exchange` object.